PFR/MeasureReal.lean

```diff
@@ -0,0 +1,378 @@
+      Finset.sum_insert]
+    exact (measureReal_union_le _ _).trans (by gcongr)
+
+theorem measureReal_iUnion_fintype_le [Fintype β] (f : β → Set α) :
+    μ.real (⋃ b, f b) ≤ ∑ p, μ.real (f p) := by
+  convert measureReal_biUnion_finset_le Finset.univ f
+  simp
+
+theorem measureReal_union_null (h₁ : μ.real s₁ = 0) (h₂ : μ.real s₂ = 0) :
+    μ.real (s₁ ∪ s₂) = 0 := by
+  apply le_antisymm _ measureReal_nonneg
+  exact (measureReal_union_le s₁ s₂).trans (by simp [h₁, h₂])
+
+@[simp]
+theorem measureReal_union_null_iff
+    (h₁ : μ s₁ ≠ ∞ := by finiteness) (h₂ : μ s₂ ≠ ∞ := by finiteness) :
+    μ.real (s₁ ∪ s₂) = 0 ↔ μ.real s₁ = 0 ∧ μ.real s₂ = 0 := by
+  have : μ (s₁ ∪ s₂) ≠ ∞ := measure_union_ne_top h₁ h₂
+  refine ⟨fun h => ⟨?_, ?_⟩, fun h => measureReal_union_null h.1 h.2⟩
+  · exact measureReal_mono_null (subset_union_left _ _) h this
+  · exact measureReal_mono_null (subset_union_right _ _) h this
+
+/-- If two sets are equal modulo a set of measure zero, then `μ.real s = μ.real t`. -/
+theorem measureReal_congr (H : s =ᵐ[μ] t) : μ.real s = μ.real t := by
+  simp [Measure.real, measure_congr H]
+
+theorem measureReal_inter_add_diff₀ (s : Set α) (ht : NullMeasurableSet t μ)
+    (h : μ s ≠ ∞ := by finiteness) :
+    μ.real (s ∩ t) + μ.real (s \ t) = μ.real s := by
+  simp only [measureReal_def]
+  rw [← ENNReal.toReal_add, measure_inter_add_diff₀ s ht]
+  · exact measure_ne_top_of_subset (inter_subset_left _ _) h
+  · exact measure_ne_top_of_subset (diff_subset _ _) h
+
+theorem measureReal_union_add_inter₀ (s : Set α) (ht : NullMeasurableSet t μ)
+    (h₁ : μ s ≠ ∞ := by finiteness) (h₂ : μ t ≠ ∞ := by finiteness) :
+    μ.real (s ∪ t) + μ.real (s ∩ t) = μ.real s + μ.real t := by
+  have : μ (s ∪ t) ≠ ∞ :=
+    ((measure_union_le _ _).trans_lt (ENNReal.add_lt_top.2 ⟨h₁.lt_top, h₂.lt_top⟩ )).ne
+  rw [← measureReal_inter_add_diff₀ (s ∪ t) ht this, Set.union_inter_cancel_right, union_diff_right,
+    ← measureReal_inter_add_diff₀ s ht h₁]
+  ac_rfl
+
+theorem measureReal_union_add_inter₀' (hs : NullMeasurableSet s μ) (t : Set α)
+    (h₁ : μ s ≠ ∞ := by finiteness) (h₂ : μ t ≠ ∞ := by finiteness) :
+    μ.real (s ∪ t) + μ.real (s ∩ t) = μ.real s + μ.real t := by
+  rw [union_comm, inter_comm, measureReal_union_add_inter₀ t hs h₂ h₁, add_comm]
+
+theorem measureReal_union₀ (ht : NullMeasurableSet t μ) (hd : AEDisjoint μ s t)
+    (h₁ : μ s ≠ ∞ := by finiteness) (h₂ : μ t ≠ ∞ := by finiteness) :
+    μ.real (s ∪ t) = μ.real s + μ.real t := by
+  simp only [Measure.real]
+  rw [measure_union₀ ht hd, ENNReal.toReal_add h₁ h₂]
+
+theorem measureReal_union₀' (hs : NullMeasurableSet s μ) (hd : AEDisjoint μ s t)
+    (h₁ : μ s ≠ ∞ := by finiteness) (h₂ : μ t ≠ ∞ := by finiteness) :
+    μ.real (s ∪ t) = μ.real s + μ.real t := by
+  rw [union_comm, measureReal_union₀ hs (AEDisjoint.symm hd) h₂ h₁, add_comm]
+
+theorem measureReal_add_measureReal_compl₀ [IsFiniteMeasure μ] {s : Set α}
```

Comment:
Is the correct design choice to state this assuming instance `[IsFiniteMeasure μ]` or is it better to assume `(h : μ univ ≠ ∞ := by finiteness)` and let the `finiteness` tactic fill in the proof in particular if an instance `[IsFiniteMeasure μ]` is around. One could argue that there are not many other ways in which the tactic could prove the hypothesis, but I wonder if there is any robustness difference between the designs...

Reply:
It is equivalent to have `IsFiniteMeasure μ` and `μ univ ≠ ∞`, but when applying lemmas the experience with autoparams is just a little bit less smooth than with typeclass inference (especially when doing rewrites) so that's why I went for `[IsFiniteMeasure μ]`. Happy to change it if we encounter situations where this creates friction!